File: 블록_게임.py
def solution(board):
    def remove_if_possible(i,j, block):
        global cnt
        if i + 2 < N and board[i+2][j] == block:
            ni = i+2
            if j+1 < N and (board[i][j+1] == block or board[ni][j+1] == block or board[ni-1][j+1] == block):
                # 1-2, 2-2, 3-2
                nj = j+1
            else:
                # 1-4. 3-4
                nj = j-1
        elif i + 2 < N and j + 1 < N and board[i+2][j+1] == block:
            # 2-4
            ni, nj = i+2, j+1
        else:
            ni = i+1
            if j+2 < N and (board[i][j+2] == block or board[ni][j+2] == block):
                # 1-1, 1-3, 2-1, 3-3
                nj = j+2
            elif 0 <= j-2 and board[ni][j-2] == block:
                # 2-3
                nj = j-2
            else:
                if board[i+1][j] == block:
                    j-=1
                    nj = j + 2
        # check
        for x in range(i, ni+1):
            if j > nj:
                for y in range(nj, j+1):
                    if board[x][y] not in [block, '*']:
                        return
            else:
                for y in range(j, nj+1):
                    if board[x][y] not in [block, '*']:
                        return

        for x in range(i, ni+1):
            if j > nj:
                for y in range(nj, j+1):
                    board[x][y] = 0
            else:
                for y in range(j, nj+1):
                    board[x][y] = 0 
        cnt += 1
        return 
    
    N = len(board)
    global cnt
    cnt = 0
    print('########### original board ##########')
    for i in range(N):
        print([str(x) for x in board[i]])
    print()

    while True:
        xcnt = cnt
        visited = []
        # 1x1 블록 뿌리기
        for j in range(N):
            for i in range(N):
                if board[i][j] != 0:
                    if 0 <= i - 2:
                        board[i-2][j] = '*'
                    if 0 <= i - 1:
                        board[i-1][j] = '*'
                    break
        print('----------------------------------------------------------')
        print('>>>> 1x1 block')
        for i in range(N):
            print([str(x) for x in board[i]])
        print()

        # 없앨 수 있는 블룩 확인하고 없애기
        for i in range(N):
            for j in range(N):
                if board[i][j] not in [0, '*'] and board[i][j] not in visited:
                    block = board[i][j]
                    visited.append(block)
                    remove_if_possible(i,j,block)

        if xcnt == cnt:
            break

        # 미사용 1x1 블록 삭제
        for i in range(N):
            for j in range(N):
                if board[i][j] == '*':
                    board[i][j] = 0

        print('>>>> After Remove')
        for i in range(N):
            print([str(x) for x in board[i]])
        print()

    return cnt

boards = [
    [[0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 4, 0, 0, 0], [0, 0, 0, 0, 0, 4, 4, 0, 0, 0], [0, 0, 0, 0, 3, 0, 4, 0, 0, 0], [0, 0, 0, 2, 3, 0, 0, 0, 5, 5], [1, 2, 2, 2, 3, 3, 0, 0, 0, 5], [1, 1, 1, 0, 0, 0, 0, 0, 0, 5]],
    [[0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0],[0,0,0,2,2,0,0,0,0,0],[0,0,0,2,1,0,0,0,0,0],[0,0,0,2,1,0,0,0,0,0], [0,0,0,0,1,1,0,0,0,0], [0,0,0,0,0,0,0,0,0,0], [0,0,0,0,0,0,0,0,0,0]]
] 
answers = [
    2,1
]

for idx in range(len(boards)):
    v = solution(boards[idx])
    if v == answers[idx]:
        print(f'[성공] 실행한 결과값 {v}')
    else:
        print(f'[실패] 실행한 결과값 {v}이(가) 기대값 {answers[idx]}와(과) 다릅니다.')

File: js/블록_게임.js
// FAIL

function solution(board) {
  const N = board.length;
  // 오른쪽, 왼쪽, 아래 순으로 탐색
  const DR = [0,  0, 1];
  const DC = [1, -1, 0];
  const checked = new Array(N);
  for (let i = 0; i < N; i++) {
    checked[i] = new Array(N).fill(0);
  }

  let cnt = 0;

  for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
      // 블록 번호
      const block = board[i][j];
      // 방문할 필요가 없을 경우 패스
      if (checked[i][j]) continue;
      // 빈 곳인 경우 
      if (block === 0) continue;

      // i, j는 방문처리
      checked[i][j] = 1;
      const blockList = [[i, j]];
      const Q = [[i, j]];
      while (Q.length) {
        const [r, c] = Q.shift();
        for (let d = 0; d < 3; d++) {
          const nr = r + DR[d];
          const nc = c + DC[d];
          // 경계 확인
          if (nr < 0 || nc < 0 || nr >= N || nc >= N) continue;
          // 체크 확인
          if (checked[nr][nc]) continue;
          // 번호 확인
          if (board[nr][nc] !== block) continue;

          blockList.push([nr, nc])
          Q.push([nr, nc])
          checked[nr][nc] = 1;
        }
      }

      // blockList가 4개 미만일 경우 이미 위에서 막힌 경우
      if (blockList.length < 4) {
        blockList.forEach(([r, c]) => {
          while (r < N) {
            checked[r][c] = 1;
            r += 1;
          }
        })
        continue
      };

      // 해당 블록 리스트가 제거 가능한 것인지 확인
      /* 
          x    |     x
          x    |     x
          x x  |   x x
      */
      if (blockList[0][1] === blockList[1][1] && blockList[1][1] === blockList[2][1]) {
        const r1 = blockList[0][0];
        const r2 = blockList[1][0];
        const c = blockList[3][1];
        if (board[r1][c] === 0 && board[r2][c] === 0) {
          console.log(block);
          cnt += 1;
          continue;
        }
      } 

      /*  
            x     |   x       |       x
          x x x   |   x x x   |   x x x
      */
      if (blockList[1][0] === blockList[2][0] && blockList[2][0] === blockList[3][0]) {
        const r = blockList[0][0];
        const c1 = blockList[2][1];
        const c2 = blockList[3][1];
        if (board[r][c1] === 0 && board[r][c2] === 0) {
          blockList.forEach(([r, c]) => {
            board[r][c] = 0;
          })
          cnt += 1;
          continue;
        }
      }

      // 해당 블록의 아래에 있는 것들은 모두 checked = 1로 변경
      blockList.forEach(([r, c]) => {
        while (r < N) {
          blockList.forEach(([r, c]) => {
            board[r][c] = 0;
          })
          checked[r][c] = 1;
          r += 1;
        }
      })
    }
  }

  return cnt;
}

const board = [
  [0,0,0,0,0,0,0,0,0,0],
  [0,0,0,0,0,0,0,0,0,0],
  [0,0,0,0,0,0,0,0,0,0],
  [0,0,0,0,0,0,0,0,0,0],
  [0,0,0,0,0,0,4,0,0,0],
  [0,0,0,0,0,4,4,0,0,0],
  [0,0,0,0,3,0,4,0,0,0],
  [0,0,0,2,3,0,0,0,5,5],
  [1,2,2,2,3,3,0,0,0,5],
  [1,1,1,0,0,0,0,0,0,5]
]
// const board = [
//   [0,0,0,0,0,0,0,0,0,0],
//   [0,0,0,0,0,0,0,0,0,0],
//   [0,0,0,0,0,0,0,0,0,0],
//   [0,0,0,0,0,0,0,0,0,0],
//   [0,0,0,0,0,0,0,0,0,0],
//   [0,0,0,0,0,0,0,0,0,0],
//   [0,0,0,0,0,0,1,0,0,0],
//   [0,0,0,2,2,2,1,0,0,0],
//   [0,0,0,2,0,1,1,0,0,0],
//   [0,0,0,0,0,0,0,0,0,0]
// ]
console.log(solution(board))